### PR TITLE
js: lift status state from edge and edgegroup to Graph

### PR DESCRIPTION
**Pull request by leowrites**
Moved all state from EdgeGroup that tracks if an edge missing to Graph and passed in the states as a prop to EdgeGroup, and a callback function to update the state. 
Moved individual edge status from Edge to Graph. Passed down the states to Edge as props and a callback for updating that edge.

## Motivation and Context
This pull request lifted states for Edge and EdgeGroup, so in the future we can implement a state management architecture.

## Your Changes
Created a new edgeMissingStatus state in Graph.
Created a new function updateEdgeMissingStatus in Graph to update if an Edge is "missing" and passed it to EdgeGroup.
EdgeGroup now calls update EdgeMissingStatus to update if an edge is missing.
Created a new edgesStatus state in Graph.
Created a new function updateEdgeStatus in Graph to update the state of edgeStatus. Passed updateEdgeStatus to Edge to update its status in Graph.
Modified code in Node and Bool for updateEdgeStatus to work.

**Description**:

**Type of change** (select all that apply):

- [x] Refactoring (internal change to codebase, without changing functionality)

## Testing

`yarn test` was ran to make sure all components are still working as expected.

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the CircleCI tests have passed.

### CODE REVIEW

File: js/components/graph/Graph.js [hunk 1]
```diff
@@ -33,6 +34,8 @@ export class Graph extends React.Component {
       hybridsJSON: [],
       boolsJSON: [],
       edgesJSON: [],
+      edgeMissingStatus: {},
```
Comment (david-yz-liu) [inline js/components/graph/Graph.js:37]: This status isn't actually used anywhere, and can be deleted, along with all of the `updateEdgeMissingStatus` code.

File: js/components/graph/EdgeGroup.js [hunk 2]
```diff
@@ -19,28 +15,20 @@ export default class EdgeGroup extends React.Component {
...
-    this.props.edgesJSON.forEach(edgeJSON => {
-      this[edgeJSON.id_].updateStatus()
-    })
+    this.props.updateEdgeMissingStatus(edgeID, isMissing)
   }
 
   /**
    * Set all Edges to the inactive status
    */
   reset = () => {
```
Comment (david-yz-liu) [inline js/components/graph/EdgeGroup.js:26]: Instead of modifying this method, you should move the corresponding logic into the `Graph.reset` method so that it can reset the edge status state itself (rather than this indirect approach of `Graph.reset` calling `EdgeGroup.reset`.

This will also allow you to delete the `ref` that the `Graph` has to the `EdgeGroup` child component.

File: js/components/graph/EdgeGroup.js [hunk 3]
```diff
@@ -7,10 +7,6 @@ import Edge from "./Edge"
 export default class EdgeGroup extends React.Component {
   constructor(props) {
     super(props)
-    // EdgeGroup's state is used to keep track of the edgeIDs of
```
Comment (david-yz-liu) [inline js/components/graph/EdgeGroup.js:10]: This is good, but now that you've removed all of the state there's no reason to override the constructor at all, so you can just delete this entire method.

File: js/components/graph/Edge.js [hunk 4]
```diff
@@ -72,5 +47,6 @@ Edge.propTypes = {
   /** Node that the edge is pointing to */
   target: PropTypes.string,
   /** function called when the edge's state has changed */
+  edgeStatus: PropTypes.string,
```
Comment (david-yz-liu) [inline js/components/graph/Edge.js:50]: Name this prop `status` instead of `edgeStatus`. It's not necessary to specify "edge" here since that's already the name of this component class.

File: js/components/graph/Node.js [hunk 5]
```diff
@@ -200,7 +214,7 @@ export default class Node extends React.Component {
     })
     this.props.inEdges.forEach(function (edge) {
       var currentEdge = svg.edges.current[edge]
-      if (currentEdge.state.status === "missing") {
+      if (currentEdge.props.status === "missing") {
```
Comment (david-yz-liu) [inline js/components/graph/Node.js:217]: As a comment, this line couldn't have been working properly before because you had named the prop `edgeStatus` instead of status. This is a good example of a case where tests should catch an error.

File: js/components/graph/Graph.js [hunk 6]
```diff
@@ -285,6 +293,45 @@ export class Graph extends React.Component {
...
+    })
+  }
+
+  /**
+   * Update the status of the Edge, this will take in a source, a target so it can be looked up
+   * We will loop through the edgesJSON and find use the source, target directly from edgeJSON
+   * the state will be stored and updated in edgeGroup object.
+   */
+  updateEdgeStatus = (source, target, status, edgeID) => {
+    const sourceNode = refLookUp(source, this)
```
Comment (david-yz-liu) [inline js/components/graph/Graph.js:310]: Move this logic (up to line 325) into the `Edge.updateStatus` function. The `Graph.updateEdgeStatus` function should only take in the `status` and `edgeID` to update.

**Review by david-yz-liu — COMMENTED** [review]
@leowrites good start. In addition to addressing the comments I left, please work on adding tests to this PR.

File: js/components/graph/Edge.js [hunk 7]
```diff
@@ -8,46 +8,33 @@ import { refLookUp } from "../common/utils"
 export default class Edge extends React.Component {
   constructor(props) {
     super(props)
-    this.state = { status: "inactive" }
```
Comment (david-yz-liu) [inline js/components/graph/Edge.js:11]: Now that you've deleted the only state, you don't need to override the constructor at all.

File: js/components/graph/Edge.js [hunk 8]
```diff
@@ -8,46 +8,33 @@ import { refLookUp } from "../common/utils"
...
+        status = "inactive"
+      } else if (!targetNode.isSelected()) {
+        status = "takeable"
+      } else {
+        status = "active"
+      }
     }
+    this.props.updateEdgeStatus(status, this.props.edgeID)
   }
-
```
Comment (david-yz-liu) [inline js/components/graph/Edge.js:50]: Revert this change

File: js/components/graph/Graph.js [hunk 9]
```diff
@@ -791,7 +817,7 @@ export class Graph extends React.Component {
             var currentEdge = svg.edges.current[edge]
             var sourceNode = refLookUp(currentEdge.props.source, svg)
             if (!sourceNode.isSelected()) {
-              currentEdge.setState({ status: "missing" })
+              currentEdge.updateStatus("missing")
```
Comment (david-yz-liu) [inline js/components/graph/Graph.js:820]: Because this function is now in `Graph`, we shouldn't need to call `currentEdge.updateStatus` with a specific argument. Instead, you can call the `updateEdgeStatus` function directly here.

File: js/components/graph/Graph.js [hunk 10]
```diff
@@ -1038,7 +1060,13 @@ export class Graph extends React.Component {
             updateNode={this.updateNode}
             focusPrereqs={this.focusPrereqs}
           />
-          <EdgeGroup svg={this} ref={this.edges} edgesJSON={this.state.edgesJSON} />
+          <EdgeGroup
+            svg={this}
+            ref={this.edges}
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply js/components/graph/Graph.js:1069]: Based on our conversation today, you can keep this for now but aim to remove it in the future.

File: js/components/graph/Edge.js (hunk 8 above)
Comment (leowrites) [reply js/components/graph/Edge.js:50]: Should I move the logic from line 18 to 37 into Graph like we discussed?

Comment (david-yz-liu) [reply js/components/graph/Edge.js:50]: @leowrites you don't need to do that for this PR. It's better to get this one merged in first, and then do subsequent updates in the Node PR.

**Review by david-yz-liu — APPROVED** [review]
